### ISSUE TITLE
[discussion] General

### ISSUE THREAD

**Issue by mathiask88** [kkoopa/1to2#1]
Let's move the discussion from https://github.com/nodejs/nan/pull/364 to this place. I created an empty electron app to serve as a basis for this tool.
![unbenannt](https://cloud.githubusercontent.com/assets/4368785/7931733/53322912-0913-11e5-8281-e25adbb3ba79.png)

Next I will look into parsing `binding.gyp` files to feed your conversion script. I don't think this is so much work, because every target got `sources` and `include_dirs`.

Including `libclang` will require a copyright notice http://llvm.org/releases/3.6.1/LICENSE.TXT. I think we should serve a small licence page in the app?

Then we need to prebuild and pack everything for Linux/Win/Mac and we are done ;)

**Comment by kkoopa** [comment]
Cool. binding.gyp should be json or something.

**Comment by mathiask88** [comment]
Yup, json. I forgot one point, libclang needs node sources. Should the user provide that path or should we download or bundle a specific version?

**Comment by kkoopa** [comment]
Can't bundle. Need the correct versions. And it only needs the headers. Could download some version, like node-gyp does.

**Comment by mathiask88** [comment]
Downloading is good, but what is the correct version?

**Comment by kkoopa** [comment]
The currently executing version, unless overridden.

**Comment by mathiask88** [comment]
What if the user wants to migrate an older project with e.g NAN 1.1? Maybe we should use the oldest node version headers NAN 1.x supports?

Edit: Or a version depending on the NAN version in package.json

**Comment by kkoopa** [comment]
Not going to support that. If it works by accident, good. Else, upgrade to NAN 1.9 first.

**Comment by mathiask88** [comment]
Ah ok, so a 1.9to2 script, so I can show a warning for <1.9. But I don't get why it must be be the current executing version. We need the headers that libclang can resolve everything, so 0.10.36 headers are as good as 2.0? Sorry for buggin you, but bundling the headers is much less work.

**Comment by kkoopa** [comment]
It will probably not work, or lead to strange problems. It should use the correct, supplied headers.

**Comment by mathiask88** [comment]
Can you explain that to me so I understand it as well? BTW I noticed GYP files are not valid JSON they are python. But a YAML parser does the job quiet well. What do you need from the gyp files (binding.gyp, common.gypi,...)? Filenames, include dirs, preprocessor definitions? If you give me write access to this repo I can upload what I got so far into a WIP branch.

**Comment by kkoopa** [comment]
The difference between versions are many. io.js 2.0 uses V8 4.2, which uses C++11 features, so that flag has to be enabled there. Different types of things between V8 / node versions, etc. I don't know exactly what problems it may or may not cause. To make it easy, just make it possible to give a folder with the necessary headers, which have been manually downloaded and extracted. It can always be improved later.

Filenames, include dirs and preprocessor definitions and such, all of them are command line arguments to the compiler and all which work with clang should ideally be included. Again, to make it easier, just allow specifying args manually and do possible automation afterwards.

I've added you as collaborator to this repository.

**Comment by mathiask88** [comment]
My thoughts were that if I configure an addon with node-gyp and any target version supported by nan the code will compile and if it compiles the parser we use is happy.

Ok, will do as you want.

Edit: These buttons are really too close....

**Comment by kkoopa** [comment]
It ought to work, but I can't really tell in advance. So many things to keep track of simultaneously.

**Comment by kkoopa** [comment]
So, I've discovered one concrete reason: Conditional compilation. The tool needs to run under each relevant set of `#ifdef`s and `#if`s. Nothing can be done about that.

**Comment by mathiask88** [comment]
Yes you are right. Not the fastest approach but if the results are good let's do this and scale the jobs across CPUs. I think most defines are already in the binding.gyp or common.gypi.

**Comment by kkoopa** [comment]
Also needs to run on different platforms if there is conditional compilation for them, e.g. Windows and Linux.

Further, it seems like multiple passes will be necessary to do all the replacements correctly, or there will be a need for a much more advanced patching system than what I currently have. I'll still try and look into it if I can come up with something better. Perhaps some delayed execution will work.

In the end, this tool will not be perfect either. Some problems are easier than with regular expressions, some are harder. I'm aiming at doing the easy 80 % automatically, leaving less actual work for the developer.

I think I finally have all the necessary infrastructure in place regarding the libclang bindings. Now the remaining work should mostly be about actually doing the replacements, which was the original task. Libclang 3.6 or newer is required, as older versions don't support template argument type extraction.

**Comment by mathiask88** [comment]
> Also needs to run on different platforms if there is conditional compilation for them, e.g. Windows and Linux.

Ok, this is bad :/ Libclang 3.6, as long as the windows dll got every function (I will test this), is good. I wanted to bundle it anyway. Is it possible  to bundle OS headers so that the user doesn't need 3 OSs(VMs)?

**Comment by kkoopa** [comment]
In theory, I suppose, yes, in practice no. This is the users' problem. If they 
care about supporting many platforms, it can be assumed they have the ability 
to test on said platforms. Those machines could be used to do the conversion.
